src/ResetMap.jl

```diff
@@ -0,0 +1,169 @@
+import Base: isempty
+
+export ResetMap,
+       apply,
+       apply!
+
+"""
+    ResetMap{N<:Real} <: LazySet{N}
+
+Type that represents a lazy reset map.
+A reset map is a special case of an affine map ``Ax + b`` where the linear map
+``A`` is the identity matrix with zero entries in all reset dimensions, and the
+translation ``b`` is zero in all non-reset dimensions.
+
+### Fields
+
+- `X`      -- convex set
+- `resets` -- resets (a mapping from an index to a new value)
+
+### Example
+
+```jldoctest resetmap
+julia> X = BallInf([2.0, 2.0, 2.0], 1.0);
+
+julia> r = Dict(1 => 4.0, 3 => 0.0);
+
+julia> rm = ResetMap(X, r);
+
+```
+
+Here `rm` modifies the set `X` such that `x1` is reset to 1 and `x3` is reset to
+0, while `x2` is not modified.
+The corresponding affine map ``A x + b`` would be
+
+```math
+\begin{pmatrix} 0 & 0 & 0 \\ 0 & 1 & 0 \\ 0 & 0 & 0 \end{pmatrix} x +
+\begin{pmatrix} 4 & 0 & 0 \end{pmatrix}
+```
+
+Applying the reset map to a vector `x = [1.0, 2.0, 3.0]`, we get:
+
+```jldoctest resetmap
+julia> apply(rm, [1.0, 2.0, 3.0])
+3-element Array{Float64,1}:
+ 4.0
+ 2.0
+ 0.0
+```
+"""
+struct ResetMap{N<:Real} <: LazySet{N}
+    X::LazySet{N}
+    resets::Dict{Int, N}
+end
+
+"""
+    apply(rm::ResetMap{N}, x::AbstractVector{N}) where {N<:Real}
+
+Apply a reset map to a given vector.
+
+### Input
+
+- `rm` -- reset map
+- `x`  -- vector
+
+### Output
+
+A fresh vector corresponding to `x` after `rm` was applied.
+"""
+function apply(rm::ResetMap{N}, x::AbstractVector{N}) where {N<:Real}
+    return apply!(rm, copy(x))
+end
+
+"""
+    apply!(rm::ResetMap{N}, x::AbstractVector{N}) where {N<:Real}
+
+Apply a reset map to a given vector.
+
+### Input
+
+- `rm` -- reset map
+- `x`  -- vector (modified in this function)
+
+### Output
+
+The same (but see the Notes below) vector `x` but after `rm` was applied.
+
+### Notes
+
+The vector `x` is modified in-place if it has type `Vector` or `SparseVector`.
+Otherwise, we first create a new `Vector` from it.
+"""
+function apply!(rm::ResetMap{N}, x::AbstractVector{N}) where {N<:Real}
+    return substitute!(rm.resets, x)
```

Comment:
I think that this function should check for set membership `x \in rm.X`.

Reply:
Also, why call it `apply` / `apply!`? In this library we would call this the concrete reset map, `reset_map(...)`. It may turn out to be interesting to add an `apply` function of a lazy set, why not, but that's another discussion.

Reply:
> This is not the semantics of `ResetMap`.

Hmm but why? Otherwise, should we clarify in the docstring that it is not checked if `x` belongs to `X`?

Reply:
It is not a `ConstrainedResetMap` as in `MathematicalSystems`. The represented set is
`A x + b, x ∈ X`.

Reply:
Okay, I see now the confusion...